plugins/logs/plugin.go

```diff
@@ -401,12 +401,48 @@ func (p *Plugin) Start(ctx context.Context) error {
 // Stop stops the plugin.
 func (p *Plugin) Stop(ctx context.Context) {
 	p.logInfo("Stopping decision logger.")
+
+	gracefulDeadline, _ := ctx.Deadline()
```

Comment:
What would happen if we didn't do **any of this**, but let the passed-in `ctx`'s deadline determine what happens?

I'm thinking of something like
```go
func (p *Plugin) Stop(ctx context.Context) {
	if p.config.Service != "" {
		p.flushDecisions(ctx)
	}
	// ...
}

func (p *Plugin) flushDecisions(ctx context.Context) {
		for {
			ok, err := p.oneShot(ctx)
			if err != nil {
				p.logError("%v.", err)
			} else if ok {
				p.logInfo("All decisions in buffer uploaded.")
				return // no need to cancel this context, we're done here
			}
			// Wait some before retrying, but skip incrementing interval since we are shutting down
			time.Sleep(1 * time.Second)
		}
}
```

I may be missing something, but I think the simpler, the better, here. 😄 

There's two things that can happen:
- the drain loop in `flushDecisions` finishes flushing the buffer before the ctx timeout, in which case it'll return, and so will `Stop(ctx)`
- the loop doesn't finish, so it'll eventually call `oneShot(ctx)` with a cancelled context (because the deadline has been reached), and return an error. Whatever's running will soon be killed because the process stops.

(⚠️ I've dropped the goroutine spawning, we can keep that if we don't want `Stop(ctx)` to block. It will not have consequences wrt what I'm proposing here.)

Reply:
This is indeed a simpler model, and I'm all for simple :) Trying this change and awaiting a timeout I see it hangs (or rather repeats each second) on the error condition though:

```
time="2020-10-06T09:53:03+02:00" level=error msg="Log upload failed: Post \"http://127.0.0.1:61166/logs\": context deadline exceeded." plugin=decision_logs
time="2020-10-06T09:53:04+02:00" level=error msg="Log upload failed: Post \"http://127.0.0.1:61166/logs\": context deadline exceeded." plugin=decision_logs
time="2020-10-06T09:53:05+02:00" level=error msg="Log upload failed: Post \"http://127.0.0.1:61166/logs\": context deadline exceeded." plugin=decision_logs
```

So if no longer listening for `.Done()` we'd need to implement custom error checking here to see if the error is due to the context having been canceled (and if so, report and return) or if it's due to some other condition (and if so, continue). I'm not sure it's _that_ much simpler.

Reply:
Yeah, it's got to happen somewhere. I'm not standing in your way here, this looks decent and is working better than before 😃

Reply:
Au contraire @srenatus , I'm _very_ happy to have you standing in my way 😄 Learning a lot from you in these discussions 👍